File: liblathe/base/quadtree.py
from liblathe.base.point import Point

class Node:
    """A node centred at (cx, cy) with width w and height h."""

    def __init__(self, center = Point(), w=1, h=1):
        self.center = center
        self.w = w 
        self.h = h
        '''
        self.north_edge = cy - h/2
        self.east_edge =  cx + w/2
        self.south_edge =  cy + h/2
        self.west_edge = cx - w/2
        '''
class Quadtree:
    def __init__(self, node, segment_group, img, precision=0.1, target=0.25, limit=2, depth=0):
        self.target = target
        self.limit = limit
        self.precision = precision
        self.node = node
        self.signed_value = None
        self.divided = False
        self.segment_group = segment_group
        self.depth = depth

        self.img = img

        self.conquer()

    def conquer(self):
        """Divide each node until the target precision is reached"""
        # get signed value if less than limit, divide and conquer
        # until a the precision limit is reached
        self.signed_value = self.segment_group.sdv(self.node.center)
        # print('depth', self.depth, 'node', self.signed_value, 'w', self.node.w, 'h', self.node.h, 'x', self.node.center.X , 'z', self.node.center.Z)
        
        self.draw()

        # turning parameters
        stepover = 0.25
        passes = 5

        offset = stepover * passes * 1.5

        if self.divided:
            return
        
        if self.node.w <= 0.2:
            return
        
        if self.depth >= 3 and abs(self.signed_value) > offset * 3:
            return
        
        if self.depth < 6 or self.signed_value > 0 and self.signed_value < offset * 2.5:
            self.divide()

        if self.signed_value < 0 and abs(self.signed_value) < offset:
            self.divide()

        '''
        if self.depth >= 4 and abs(self.signed_value) > offset * 2:
            return

        if self.node.h <= 0.01:
            return

        modulo = 0.03
        if self.depth < 4 or (stepover - (self.signed_value)) % stepover < modulo or (self.signed_value) % stepover < modulo:            
            self.divide()
        '''


    def draw(self):

        tz = ((self.node.center.Z - self.node.w / 2) + 60 )* 10
        tx = (self.node.center.X - self.node.h / 2) * 10
        shape = [(tz, tx), (tz + self.node.w * 10, tx + self.node.h * 10)]
        colour = "#ffff33"
        if self.signed_value < 0:
            colour = "#00ffff"
        self.img.rectangle(shape, outline = colour)


    def divide(self):
        """Divide (branch) this node by spawning four children nodes."""

        cx = self.node.center.X
        cy = self.node.center.Z
        w = self.node.w / 2
        h = self.node.h / 2

        self.ne = Quadtree(Node(Point(cx - h/2, 0, cy + w/2), w, h),
                                    self.segment_group, self.img, self.precision, self.target, self.limit, self.depth+1)
        self.se = Quadtree(Node(Point(cx + h/2, 0, cy + w/2), w, h),
                                    self.segment_group, self.img, self.precision, self.target, self.limit, self.depth+1)
        self.sw = Quadtree(Node(Point(cx + h/2, 0, cy - w/2), w, h),
                                    self.segment_group, self.img, self.precision, self.target, self.limit, self.depth+1)
        self.nw = Quadtree(Node(Point(cx - h/2, 0, cy - w/2), w, h),
                                    self.segment_group, self.img, self.precision, self.target, self.limit, self.depth+1)

        self.divided = True

    def query(self, target, found_points=[]):
        """Find the points in the quadtree that are close to target value"""

        dist = round(self.signed_value, 2)
        if dist >= target and dist <= target + 0.05:
            found_points.append(self.node.center)

        if self.divided:
            self.nw.query(target, found_points)
            self.ne.query(target, found_points)
            self.se.query(target, found_points)
            self.sw.query(target, found_points)

        return found_points

File: liblathe/base/quadtree.cpp
#include <iostream>
#include <limits>

#include "quadtree.h"

Quadtree::Quadtree()
{
}

Quadtree::~Quadtree() {}

void Quadtree::initialise(SegmentGroup *segmentgroup, Point center, float width, float height)
{
    // Initialise the quadtree //
    // depth is the current node depth
    // basenode the the primary tree node

    this->segmentGroup = segmentgroup;

    // check width and height are the same
    // TODO: remove the width and height if they need to match
    if (width != height)
    {
        throw std::runtime_error("quadtree width and height must match");
    }
    // create the base node
    int depth = 0;
    Node bn = {center, width, height, depth};
    this->basenode = bn;

    // set the precision
    this->precision = 0.2;
}

std::vector<Point> Quadtree::getOffset(float offsetValue)
{
    // return the points that represent the calculated offset //
    this->offset = offsetValue;
    // start to process the quadtree, populating the offsetBoundaryPoints
    this->conquer(this->basenode);
    // sort the points so they can be used to build a tool path
    return this->sortPoints(segmentGroup->getSegments()[0].start, this->offsetBoundaryPoints);
}

void Quadtree::conquer(Node &node)
{
    // Recursively divide each node until the target precision is reached //

    // get the nodes signed distance value
    // representing the distance from the closest point on the segmentGroup
    // negative values are inside the segmentGroup
    // positive values are outside the segmentGroup
    node.sdv = this->segmentGroup->sdv(node.center);

    // check if it is possible that this node contains the target offset
    if (this->nodeCouldContain(this->offset, node))
    {
        // check if we have reached the desired precision
        if ((node.width <= this->precision))
        {
            // check that the signed distance value is the same sign as the requested offset
            if (std::signbit(node.sdv) == std::signbit(this->offset))
            {
                // get the closest point on the segmentGroup
                Point closest = this->segmentGroup->closestPoint(node.center);
                // get the projected boundary point
                Point projectedPoint = closest.project(closest.angleTo(node.center), abs(this->offset));

                // check if projected point is inside the node
                if (this->insideNode(projectedPoint, node))
                {
                    Point segmentStart = segmentGroup->getSegments()[0].start;
                    // only collect points inside the segment boundary
                    if (projectedPoint.Z <= segmentStart.Z)
                    {
                        this->offsetBoundaryPoints.push_back(projectedPoint);
                    }
                }
            }

            return;
        }
        else
        {
            this->divide(node);
        }
    }
}

void Quadtree::divide(Node &node)
{
    // Divide this node by creating four child nodes //

    float cx = node.center.X;
    float cy = node.center.Z;
    float w = node.width / 2;
    float h = node.height / 2;
    int depth = node.depth + 1;

    float x = cx - h / 2;
    float z = cy + w / 2;
    Point pne = Point(x, z);
    Node ne = {pne, w, h, depth};

    x = cx + h / 2;
    z = cy + w / 2;
    Point pse = Point(x, z);
    Node se = {pse, w, h, depth};

    x = cx + h / 2;
    z = cy - w / 2;
    Point psw = Point(x, z);
    Node sw = {psw, w, h, depth};

    x = cx - h / 2;
    z = cy - w / 2;
    Point pnw = Point(x, z);
    Node nw = {pnw, w, h, depth};

    node.child_nodes.push_back(ne);
    node.child_nodes.push_back(se);
    node.child_nodes.push_back(sw);
    node.child_nodes.push_back(nw);

    node.divided = true;

    for (auto &child : node.child_nodes)
    {
        this->conquer(child);
    }
}

bool Quadtree::nodeCouldContain(float offset, Node &node)
{
    // check if the target offset could be within the node//

    // get the distance from the node center to a corner of the node
    // Based on pythagoras theorem: A^2 + B^2 = C^2; assumes the node is square
    // This is a quick check and could include false positives
    float cornerDistance = sqrt(pow(node.width, 2) * 2) * 0.5;

    if (node.sdv - cornerDistance <= offset && node.sdv + cornerDistance >= offset)
    {
        return true;
    }

    return false;
}

bool Quadtree::insideNode(Point &point, Node &node)
{
    // determine if point is inside node //
    float nodeXMin = node.center.X - node.height * 0.5;
    float nodeXMax = node.center.X + node.height * 0.5;
    float nodeZMin = node.center.Z - node.width * 0.5;
    float nodeZMax = node.center.Z + node.width * 0.5;

    // point is inside if it lies on the min boundary of node
    if (point.X >= nodeXMin && point.X < nodeXMax)
    {
        if (point.Z >= nodeZMin && point.Z < nodeZMax)
            return true;
    }

    return false;
}

std::vector<Point> Quadtree::sortPoints(Point datum, std::vector<Point> &points)
{
    // sort the point set into a ordered set of points starting from datum //
    std::vector<Point> sorted_points;
    int point_count = points.size();
    Point target = datum;
    int closest_index;

    int input_point_count = points.size();

    while (points.size() != 0)
    {
        float dist = std::numeric_limits<float>::infinity();
        int index = 0;

        for (index; index < points.size(); index++)
        {
            // find closest point
            float target_to_point = target.distanceTo(points[index]);

            if (target_to_point < dist)
            {
                closest_index = index;
                dist = target_to_point;
            }
        }

        // add closest point to sorted points
        sorted_points.push_back(Point(points[closest_index].X, points[closest_index].Z));
        // remove point from points array
        points.erase(points.begin() + closest_index);
        // set target to last found point
        target = sorted_points.back();
    }

    if (input_point_count != sorted_points.size())
    {
        throw std::runtime_error("Quadtree error when ordering offset points");
    }
    return sorted_points;
}

std::vector<Point> Quadtree::query(Node &node, std::vector<Point> &offsetBoundaryPoints)
{
    // Find the points in the quadtree that are close to target value //

    float dist = node.sdv;
    if (dist - this->precision * 0.5 >= this->offset && dist + this->precision * 0.5 <= this->offset)
    {
        offsetBoundaryPoints.push_back(node.center);
    }

    if (node.divided)
    {
        for (auto &child : node.child_nodes)
        {
            this->query(child, offsetBoundaryPoints);
        }
    }

    return offsetBoundaryPoints;
}

std::vector<Node> Quadtree::getNodes()
{
    // return list of nodes //
    std::vector<Node> nodes;
    return this->queryNodes(this->basenode, nodes);
}

std::vector<Node> Quadtree::queryNodes(Node &node, std::vector<Node> &nodes)
{
    // build list of nodes //
    nodes.push_back(node);

    if (node.divided)
    {
        for (auto &child : node.child_nodes)
        {
            this->queryNodes(child, nodes);
        }
    }

    return nodes;
}
